cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs

```diff
@@ -353,10 +378,301 @@ runQueryUTxO (AnyConsensusModeParams cModeParams)
+                                 (fromIntegral $ kesIntervalEnd * fromIntegral protocolParamSlotsPerKESPeriod)
+      -- We get the operational certificate counter from the protocol state and check that
+      -- it is equivalent to what we have on disk.
+
+      let ptclStateQinMode = QueryInEra eInMode . QueryInShelleyBasedEra sbe $ QueryProtocolState
+      ptclState <- executeQuery era cModeParams localNodeConnInfo ptclStateQinMode
+      (opCertCounter, ptclStateCounter, opCertCounterStateDiag) <- opCertCounterStateCheck ptclState opCert
+
+      let diagnoses = opCertCounterStateDiag ++  [periodCheckDiag]
+
+      if any anyFailureDiagnostic diagnoses
+      then
+        case mOutFile of
+          Just (OutputFile _) -> liftIO $ mapM_ kesOpCertDiagnosticsRender diagnoses
+          Nothing -> liftIO $ mapM_ kesOpCertDiagnosticsRender diagnoses
+      else do
+        let kesPeriodInfo = O.QueryKesPeriodInfoOutput
+                              { O.qKesInfoCurrentKESPeriod = currentKesPeriod
+                              , O.qKesInfoStartKesInterval = kesIntervalStart
+                              , O.qKesInfoStartEndInterval = kesIntervalEnd
+                              , O.qKesInfoRemainingSlotsInPeriod = slotsTillNewKesKey
+                              , O.qKesInfoNodeStateOperationalCertNo = ptclStateCounter
+                              , O.qKesInfoOnDiskOperationalCertNo = opCertCounter
+                              , O.qKesInfoMaxKesKeyEvolutions = fromIntegral protocolParamMaxKESEvolutions
+                              , O.qKesInfoSlotsPerKesPeriod = fromIntegral protocolParamSlotsPerKESPeriod
+                              , O.qKesInfoKesKeyExpiry = kesKeyExpirySlotUtc
+                              }
+            kesPeriodInfoJSON = encodePretty kesPeriodInfo
+
+        liftIO $ mapM_ kesOpCertDiagnosticsRender diagnoses
+
+        case mOutFile of
+          Just (OutputFile oFp) ->
+            handleIOExceptT (ShelleyQueryCmdWriteFileError . FileIOError oFp)
+              $ LBS.writeFile oFp kesPeriodInfoJSON
+          Nothing -> liftIO $ LBS.putStrLn kesPeriodInfoJSON
+
+    mode -> left . ShelleyQueryCmdUnsupportedMode $ AnyConsensusMode mode
+ where
+   -- We first check the operational certificate issue counter's count is
+   -- equal to the count in the operational certificate (on disk)
+   opCertCounterOndiskCheck
+     :: OperationalCertificate
+     -> ExceptT ShelleyQueryCmdError IO ( Word64 -- Operational certificate count
+                                        , KesOpCertDiagnostic)
+   opCertCounterOndiskCheck opCert = do
+     opCertCounter
+       <- firstExceptT ShelleyQueryCmdOpCertCounterReadError
+            . newExceptT $ readFileTextEnvelope AsOperationalCertificateIssueCounter opCertCountFile
+
+     let currentOpCertCount = getOpCertCount opCert
+         -- Why 'pred'? Because the op cert issue counter tells us what the next
+         -- count should be. TODO: Double check this.
```

Comment:
Is the `TODO` part of this PR?

Reply:
I doubled checked this so I can remove it